### DISCUSSION TITLE
Allow auto-import abbreviations to be configured

### DISCUSSION THREAD

**Discussion by jakebailey** [microsoft/pylance-release#2589] [Enhancement]
_Originally posted by @cdce8p in https://github.com/microsoft/pylance-release/issues/8#issuecomment-724011845_

I was thinking about this issue recently. Like many others, I tend to have some abbreviations I use quite frequently for imports. Could adding a new setting fix it? A suggestion:
```json
[{
    "suggestion": "pd",
    "import": "import pandas as pd",
    "enabled": true,
}, {
    "suggestion": "dt",
    "import": "from datetime import datetime as dt",
    "enabled": true,
}]
```

This syntax would allow anyone to customize their auto-import suggestions, while keeping the completion list clean for everybody else. Additionally a user could use this to disable really common mistakes.
This would also address #210

**Comment by heejaechang** [comment]
listing all related issues/suggestions here

https://github.com/microsoft/pylance-release/issues/3285 (make abbreviations always show up even if packages are not installed)
https://github.com/microsoft/pylance-release/issues/210 and https://github.com/microsoft/pylance-release/issues/8#issuecomment-724011845 (make abbreviations configurable)

**Comment by heejaechang** [comment]
briefly looked at it.

a few things that need consideration

1. will we let users to customize what abbr they will use with which package? 
> ex) adding ```dt``` for ```datetime```
2. will we show abbr suggestion even when package is not installed? 
> ex) ```pd``` showing ```import pandas as pd``` in completion and code action even if ```pandas``` is not installed.
3. will we let users to customize how import looks like for abbr? 
> ex) should ```plt``` do ```from matplotlib import pyplot as plt``` or ```import matplotlib.pyplot as plt```?
4. how will we deal with existing import statements?
* If we let users provide ```string``` such as ```from datetime import datetime as dt``` as what to insert as the description above, should we support merging with existing statement? ex) ```from datetime import date``` or should we even check import statement already exist? ```import pandas``` or ```from datetime import datetime``` 
* If we support, just string check? currently we do see decl (location of installed package) and check they point to the same package.
5. should we support ```enabled``` as the description above? sounds like some want to disable something like ```from turtle import pd``` from showing up in auto/add import suggestions when typing ```pd```?

* there are other consideration we currently do such as if file has ```import pandas``` already, we won't show anything for ```pd```, or if user already has ```import pandas as pd``` and if user writes ```pandas```, we will automatically convert it to ```pd```, do we want to support this behavior only when package is actually installed? or update our code, which currently only works with package installed and use decl (location of the module), to just use string match?

**Comment by starball5** [comment]
Related on Stack Overflow: [How do define custom auto-imports for Pylance/Visual Studio Code?](https://stackoverflow.com/q/78845297/11107541)
